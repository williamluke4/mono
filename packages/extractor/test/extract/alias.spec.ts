it('raw is based on type alias declaration node', () => {
  expect(
    ctx.extract(`
      export { A }
      /**
       * foobar
       */
      type A = {}
  `)
  ).toMatchInlineSnapshot(`
    Object {
      "modules": Array [
        Object {
          "isMain": true,
          "kind": "module",
          "location": Object {
            "absoluteFilePath": "/src/a.ts",
          },
          "mainExport": null,
          "name": "a",
          "namedExports": Array [
            Object {
              "isTerm": false,
              "isType": true,
              "kind": "export",
              "name": "A",
              "type": Object {
                "kind": "typeIndexRef",
                "link": "(a).A",
              },
            },
          ],
          "path": "/",
          "tsdoc": null,
        },
      ],
      "typeIndex": Object {
        "(a).A": Object {
          "kind": "alias",
          "name": "A",
          "raw": Object {
            "nodeFullText": "/**
     * foobar
     */
    type A = {};",
            "nodeText": "type A = {};",
            "typeText": "A",
          },
          "tsdoc": Object {
            "customTags": Array [],
            "examples": Array [],
            "raw": "/**
     * foobar
     */",
            "summary": "foobar",
          },
          "type": Object {
            "kind": "object",
            "props": Array [],
            "raw": Object {
              "nodeFullText": "/**
     * foobar
     */
    type A = {};",
              "nodeText": "type A = {};",
              "typeText": "A",
            },
          },
        },
      },
    }
  `)
})

it('exported type alias of number is added to type index', () => {
  expect(
    ctx.extract(`
      export type A = 1
    `)
  ).toMatchInlineSnapshot(`
    Object {
      "modules": Array [
        Object {
          "isMain": true,
          "kind": "module",
          "location": Object {
            "absoluteFilePath": "/src/a.ts",
          },
          "mainExport": null,
          "name": "a",
          "namedExports": Array [
            Object {
              "isTerm": false,
              "isType": true,
              "kind": "export",
              "name": "A",
              "type": Object {
                "kind": "typeIndexRef",
                "link": "(a).A",
              },
            },
          ],
          "path": "/",
          "tsdoc": null,
        },
      ],
      "typeIndex": Object {
        "(a).A": Object {
          "kind": "alias",
          "name": "A",
          "raw": Object {
            "nodeFullText": "export type A = 1;",
            "nodeText": "export type A = 1;",
            "typeText": "1",
          },
          "tsdoc": null,
          "type": Object {
            "base": "number",
            "kind": "literal",
            "name": "1",
          },
        },
      },
    }
  `)
})

it('exported type alias of number via typeof is added to type index', () => {
  expect(
    ctx.extract(`
      const a = 1
      export type A = typeof a
    `)
  ).toMatchInlineSnapshot(`
    Object {
      "modules": Array [
        Object {
          "isMain": true,
          "kind": "module",
          "location": Object {
            "absoluteFilePath": "/src/a.ts",
          },
          "mainExport": null,
          "name": "a",
          "namedExports": Array [
            Object {
              "isTerm": false,
              "isType": true,
              "kind": "export",
              "name": "A",
              "type": Object {
                "kind": "typeIndexRef",
                "link": "(a).A",
              },
            },
          ],
          "path": "/",
          "tsdoc": null,
        },
      ],
      "typeIndex": Object {
        "(a).A": Object {
          "kind": "alias",
          "name": "A",
          "raw": Object {
            "nodeFullText": "
    export type A = typeof a;",
            "nodeText": "export type A = typeof a;",
            "typeText": "1",
          },
          "tsdoc": null,
          "type": Object {
            "base": "number",
            "kind": "literal",
            "name": "1",
          },
        },
      },
    }
  `)
})

// todo nice test but differs in raw, PITA to manually tweak, revisit
// it('exported type alias of number acts same as if via typeof', () => {
//   const plain = ctx.extract(`export type A = 1`)
//   const fetch = ctx.extract(`const a = 1; export type A = typeof a`)
//   expect(plain).toEqual(fetch)
// })

it('exported type alias of function is added to type index', () => {
  expect(ctx.extract(`export type A = () => {}`)).toMatchSnapshot()
})

it('exported type alias of function via typeof is added to type index', () => {
  expect(ctx.extract(`const a = () => {}; export type A = typeof a`)).toMatchSnapshot()
})

// todo nice test but differs in raw, PITA to manually tweak, revisit
// it('exported type alias of function acts same as if via typeof', () => {
//   const plain = ctx.extract(`export type A = () => {}`)
//   const fetch = ctx.extract(`const a = () => {}; export type A = typeof a`)
//   expect(plain).toEqual(fetch)
// })

it('exported type alias of object is added to type index', () => {
  expect(ctx.extract(`export type A = { b: 1 }`)).toMatchSnapshot()
})

it('exported type alias of object via typeof is added to type index', () => {
  expect(ctx.extract(`const a = { b: 1 }; export type A = typeof a`)).toMatchSnapshot()
})

it('exported type alias of boolean is added to type index', () => {
  expect(ctx.extract(`export type A = boolean`)).toMatchSnapshot()
})

it('exported type alias of boolean via typeof is added to type index', () => {
  expect(ctx.extract(`let a = Math.random() ? false : true; export type A = typeof a`)).toMatchSnapshot()
})

it('exported type alias of array is added to type index', () => {
  expect(ctx.extract(`export type A = 1[]`)).toMatchSnapshot()
})

it('exported type alias of array via typeof is added to type index', () => {
  expect(ctx.extract(`let a = [1]; export type A = typeof a`)).toMatchSnapshot()
})
