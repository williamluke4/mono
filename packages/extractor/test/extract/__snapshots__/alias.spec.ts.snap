// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exported type alias of array is added to type index 1`] = `
Object {
  "modules": Array [
    Object {
      "isMain": true,
      "kind": "module",
      "location": Object {
        "absoluteFilePath": "/src/a.ts",
      },
      "mainExport": null,
      "name": "a",
      "namedExports": Array [
        Object {
          "isTerm": false,
          "isType": true,
          "kind": "export",
          "name": "A",
          "type": Object {
            "kind": "typeIndexRef",
            "link": "(a).A",
          },
        },
      ],
      "path": "/",
      "tsdoc": null,
    },
  ],
  "typeIndex": Object {
    "(a).A": Object {
      "kind": "alias",
      "name": "A",
      "raw": Object {
        "nodeFullText": "export type A = 1[];",
        "nodeText": "export type A = 1[];",
        "typeText": "A",
      },
      "tsdoc": null,
      "type": Object {
        "innerType": Object {
          "base": "number",
          "kind": "literal",
          "name": "1",
        },
        "kind": "array",
      },
    },
  },
}
`;

exports[`exported type alias of array via typeof is added to type index 1`] = `
Object {
  "modules": Array [
    Object {
      "isMain": true,
      "kind": "module",
      "location": Object {
        "absoluteFilePath": "/src/a.ts",
      },
      "mainExport": null,
      "name": "a",
      "namedExports": Array [
        Object {
          "isTerm": false,
          "isType": true,
          "kind": "export",
          "name": "A",
          "type": Object {
            "kind": "typeIndexRef",
            "link": "(a).A",
          },
        },
      ],
      "path": "/",
      "tsdoc": null,
    },
  ],
  "typeIndex": Object {
    "(a).A": Object {
      "kind": "alias",
      "name": "A",
      "raw": Object {
        "nodeFullText": "
export type A = typeof a;",
        "nodeText": "export type A = typeof a;",
        "typeText": "number[]",
      },
      "tsdoc": null,
      "type": Object {
        "innerType": Object {
          "kind": "primitive",
          "type": "number",
        },
        "kind": "array",
      },
    },
  },
}
`;

exports[`exported type alias of boolean is added to type index 1`] = `
Object {
  "modules": Array [
    Object {
      "isMain": true,
      "kind": "module",
      "location": Object {
        "absoluteFilePath": "/src/a.ts",
      },
      "mainExport": null,
      "name": "a",
      "namedExports": Array [
        Object {
          "isTerm": false,
          "isType": true,
          "kind": "export",
          "name": "A",
          "type": Object {
            "kind": "typeIndexRef",
            "link": "(a).A",
          },
        },
      ],
      "path": "/",
      "tsdoc": null,
    },
  ],
  "typeIndex": Object {
    "(a).A": Object {
      "kind": "alias",
      "name": "A",
      "raw": Object {
        "nodeFullText": "export type A = boolean;",
        "nodeText": "export type A = boolean;",
        "typeText": "boolean",
      },
      "tsdoc": null,
      "type": Object {
        "kind": "primitive",
        "type": "boolean",
      },
    },
  },
}
`;

exports[`exported type alias of boolean via typeof is added to type index 1`] = `
Object {
  "modules": Array [
    Object {
      "isMain": true,
      "kind": "module",
      "location": Object {
        "absoluteFilePath": "/src/a.ts",
      },
      "mainExport": null,
      "name": "a",
      "namedExports": Array [
        Object {
          "isTerm": false,
          "isType": true,
          "kind": "export",
          "name": "A",
          "type": Object {
            "kind": "typeIndexRef",
            "link": "(a).A",
          },
        },
      ],
      "path": "/",
      "tsdoc": null,
    },
  ],
  "typeIndex": Object {
    "(a).A": Object {
      "kind": "alias",
      "name": "A",
      "raw": Object {
        "nodeFullText": "
export type A = typeof a;",
        "nodeText": "export type A = typeof a;",
        "typeText": "boolean",
      },
      "tsdoc": null,
      "type": Object {
        "kind": "primitive",
        "type": "boolean",
      },
    },
  },
}
`;

exports[`exported type alias of function is added to type index 1`] = `
Object {
  "modules": Array [
    Object {
      "isMain": true,
      "kind": "module",
      "location": Object {
        "absoluteFilePath": "/src/a.ts",
      },
      "mainExport": null,
      "name": "a",
      "namedExports": Array [
        Object {
          "isTerm": false,
          "isType": true,
          "kind": "export",
          "name": "A",
          "type": Object {
            "kind": "typeIndexRef",
            "link": "(a).A",
          },
        },
      ],
      "path": "/",
      "tsdoc": null,
    },
  ],
  "typeIndex": Object {
    "(a).A": Object {
      "kind": "alias",
      "name": "A",
      "raw": Object {
        "nodeFullText": "export type A = () => {};",
        "nodeText": "export type A = () => {};",
        "typeText": "A",
      },
      "tsdoc": null,
      "type": Object {
        "hasProps": false,
        "isOverloaded": false,
        "kind": "callable",
        "props": Array [],
        "raw": Object {
          "nodeFullText": "export type A = () => {};",
          "nodeText": "export type A = () => {};",
          "typeText": "A",
        },
        "sigs": Array [
          Object {
            "kind": "sig",
            "params": Array [],
            "return": Object {
              "kind": "object",
              "props": Array [],
              "raw": Object {
                "nodeFullText": "",
                "nodeText": "",
                "typeText": "{}",
              },
            },
          },
        ],
      },
    },
  },
}
`;

exports[`exported type alias of function via typeof is added to type index 1`] = `
Object {
  "modules": Array [
    Object {
      "isMain": true,
      "kind": "module",
      "location": Object {
        "absoluteFilePath": "/src/a.ts",
      },
      "mainExport": null,
      "name": "a",
      "namedExports": Array [
        Object {
          "isTerm": false,
          "isType": true,
          "kind": "export",
          "name": "A",
          "type": Object {
            "kind": "typeIndexRef",
            "link": "(a).A",
          },
        },
      ],
      "path": "/",
      "tsdoc": null,
    },
  ],
  "typeIndex": Object {
    "(a).A": Object {
      "kind": "alias",
      "name": "A",
      "raw": Object {
        "nodeFullText": "
export type A = typeof a;",
        "nodeText": "export type A = typeof a;",
        "typeText": "() => void",
      },
      "tsdoc": null,
      "type": Object {
        "hasProps": false,
        "isOverloaded": false,
        "kind": "callable",
        "props": Array [],
        "raw": Object {
          "nodeFullText": "() => {}",
          "nodeText": "() => {}",
          "typeText": "() => void",
        },
        "sigs": Array [
          Object {
            "kind": "sig",
            "params": Array [],
            "return": Object {
              "kind": "primitive",
              "type": "void",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`exported type alias of object is added to type index 1`] = `
Object {
  "modules": Array [
    Object {
      "isMain": true,
      "kind": "module",
      "location": Object {
        "absoluteFilePath": "/src/a.ts",
      },
      "mainExport": null,
      "name": "a",
      "namedExports": Array [
        Object {
          "isTerm": false,
          "isType": true,
          "kind": "export",
          "name": "A",
          "type": Object {
            "kind": "typeIndexRef",
            "link": "(a).A",
          },
        },
      ],
      "path": "/",
      "tsdoc": null,
    },
  ],
  "typeIndex": Object {
    "(a).A": Object {
      "kind": "alias",
      "name": "A",
      "raw": Object {
        "nodeFullText": "export type A = { b: 1 };",
        "nodeText": "export type A = { b: 1 };",
        "typeText": "A",
      },
      "tsdoc": null,
      "type": Object {
        "kind": "object",
        "props": Array [
          Object {
            "kind": "prop",
            "name": "b",
            "type": Object {
              "base": "number",
              "kind": "literal",
              "name": "1",
            },
          },
        ],
        "raw": Object {
          "nodeFullText": "export type A = { b: 1 };",
          "nodeText": "export type A = { b: 1 };",
          "typeText": "A",
        },
      },
    },
  },
}
`;

exports[`exported type alias of object via typeof is added to type index 1`] = `
Object {
  "modules": Array [
    Object {
      "isMain": true,
      "kind": "module",
      "location": Object {
        "absoluteFilePath": "/src/a.ts",
      },
      "mainExport": null,
      "name": "a",
      "namedExports": Array [
        Object {
          "isTerm": false,
          "isType": true,
          "kind": "export",
          "name": "A",
          "type": Object {
            "kind": "typeIndexRef",
            "link": "(a).A",
          },
        },
      ],
      "path": "/",
      "tsdoc": null,
    },
  ],
  "typeIndex": Object {
    "(a).A": Object {
      "kind": "alias",
      "name": "A",
      "raw": Object {
        "nodeFullText": "
export type A = typeof a;",
        "nodeText": "export type A = typeof a;",
        "typeText": "{ b: number; }",
      },
      "tsdoc": null,
      "type": Object {
        "kind": "object",
        "props": Array [
          Object {
            "kind": "prop",
            "name": "b",
            "type": Object {
              "kind": "primitive",
              "type": "number",
            },
          },
        ],
        "raw": Object {
          "nodeFullText": "{ b: 1 }",
          "nodeText": "{ b: 1 }",
          "typeText": "{ b: number; }",
        },
      },
    },
  },
}
`;
